### PR TITLE
Swift 4 support

### PR DESCRIPTION
**Pull request by cristianames92**
## Summary ##
**Trello card**: None

The library is now fully ported to Swift 4 along with the dependencies

## Test Plan ##
 * The framework should build successfully

### CODE REVIEW

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 86.69
:white_check_mark: Code Quality (Swiftlint): 92.0


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-core-ios/branches/swift-4-support) for more details.

File: Core/Extensions/ReactiveCocoa/Signal.swift [hunk 1]
```diff
@@ -70,7 +70,7 @@ public extension SignalProtocol where Value: OptionalProtocol {
 
 }
 
-public extension SignalProtocol where Value: ResultProtocol {
+public extension Signal where Value: ResultProtocol {
```
Comment (danielaRiesgo) [inline Core/Extensions/ReactiveCocoa/Signal.swift]: Why this change @cristianames92 ?

File: Core/Extensions/ReactiveCocoa/SignalProducer.swift [hunk 2]
```diff
@@ -71,7 +71,7 @@ public extension SignalProducerProtocol where Value: OptionalProtocol {
 
 }
 
-public extension SignalProducerProtocol where Value: ResultProtocol {
+public extension SignalProducer where Value: ResultProtocol {
```
Comment (danielaRiesgo) [inline Core/Extensions/ReactiveCocoa/SignalProducer.swift]: Same here

File: Core/Extensions/UIKit/UIColor.swift [hunk 3]
```diff
@@ -25,7 +25,7 @@ public extension UIColor {
             start = hex.startIndex
         }
         
-        var hexColor = hex.substring(from: start)
+        var hexColor = String(hex.suffix(from: start))
```
Comment (danielaRiesgo) [inline Core/Extensions/UIKit/UIColor.swift]: XCode says `substring` is deprecated and we should use partial slicing instead: `var hexColor = 
 hex[start...]`

File: CoreTests/Extensions/Foundation/BundleSpec.swift [hunk 4]
```diff
@@ -264,11 +264,11 @@ public class BundleSpec: QuickSpec {
                     
                 }
                 
-                context("when the value is not the expected type") {
+                context("when the value is not the expected type but is Boolean") {
```
Comment (danielaRiesgo) [inline CoreTests/Extensions/Foundation/BundleSpec.swift]: `when the value is not the expected type but is a compatible one`
Then add another of `when the value is not the expected type and is not a compatible one`

File: CoreTests/Info.plist [hunk 5]
```diff
@@ -25,7 +25,7 @@
 	<key>My Personal Number Key</key>
 	<integer>5</integer>
 	<key>My Personal Float Key</key>
-	<real>7.1</real>
+	<integer>7</integer>
```
Comment (danielaRiesgo) [inline CoreTests/Info.plist]: Why did you change this? This isn't float anymore so the test changes.

File: CoreTests/Resources/Resources.swift [hunk 6]
```diff
@@ -26,7 +26,7 @@ class NibLoadableTableView: UITableViewHeaderFooterView, NibLoadable {
 class NibLoadableTableViewCustom: UITableViewHeaderFooterView, NibLoadable {
     
     static var nibBundle: Bundle {
-        return Bundle(for: MBProgressHUD.self)
+        return Bundle(for: AnyDisposable.self) // Testing with a class that's outside of our framework, it could be any class from outside
```
Comment (danielaRiesgo) [inline CoreTests/Resources/Resources.swift]: Great the comment, but I would move it underneath or on top.

File: Core/Extensions/ReactiveCocoa/Signal.swift (hunk 1 above)
Comment (cristianames92) [reply Core/Extensions/ReactiveCocoa/Signal.swift]: Reactive cocoa seemed to remove it

File: Core/Extensions/ReactiveCocoa/SignalProducer.swift (hunk 2 above)
Comment (cristianames92) [reply Core/Extensions/ReactiveCocoa/SignalProducer.swift]: ditto

File: CoreTests/Info.plist (hunk 5 above)
Comment (cristianames92) [reply CoreTests/Info.plist]: Don't even remember why I did that. will take a look at it

File: Core/Extensions/ReactiveCocoa/Signal.swift (hunk 1 above)
Comment (danielaRiesgo) [reply Core/Extensions/ReactiveCocoa/Signal.swift]: I still have it :/

File: CoreTests/Info.plist (hunk 5 above)
Comment (cristianames92) [reply CoreTests/Info.plist]: https://github.com/realm/realm-cocoa/issues/5045 Here is some info about it.
In one test you are trying to load a float value as an Int, but with this new changes Apple made on the casting system, you can no longer do that without explicitly truncating the float number, so I had two options here, transforming the numeric value to be an integer, so test will pass, or deleting this conflictive tests. I choose the first option but I can easily do the other one

Comment (cristianames92) [reply CoreTests/Info.plist]: https://github.com/Wolox/wolmo-core-ios/blob/master/CoreTests/Extensions/Foundation/BundleSpec.swift?utf8=%E2%9C%93#L259 Here is the tests I was talking about

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 86.69
:white_check_mark: Code Quality (Swiftlint): 92.0


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-core-ios/branches/swift-4-support) for more details.

File: CoreTests/Extensions/Foundation/BundleSpec.swift [hunk 7]
```diff
@@ -264,14 +264,23 @@ public class BundleSpec: QuickSpec {
                     
                 }
                 
-                context("when the value is not the expected type") {
+                context("when the value is not the expected type but it is a compatible one") {
                     
-                    it("should return .none") {
+                    it("should return numeric representation") {
```
Comment (danielaRiesgo) [inline CoreTests/Extensions/Foundation/BundleSpec.swift]: I think this test is the same as what was tried to do with `context("when the value can be converted to expected type"), it("should return the value")`. So we can erase that one and leave this one.
If `"My Personal Float Key"` is not used anywhere else, then you can just delete it from Info.plist.
Also, personally I would change the description of the it to `should return the value converted` (since you say numeric because you are trying to cast to numeric in this particular test, but it represents any possible casting case)

File: CoreTests/Info.plist (hunk 5 above)
Comment (danielaRiesgo) [reply CoreTests/Info.plist]: 👍 in another comment left why we can erase this :)

**Comment by woloxci** [comment]
**CodeStats metrics details:**

:white_check_mark: Code Coverage (Slather): 86.69
:white_check_mark: Code Quality (Swiftlint): 92.0


Click [here](http://codestats.wolox.com.ar/organizations/wolox/projects/wolmo-core-ios/branches/swift-4-support) for more details.

**Comment by cristianames92** [comment]
I cannot merge this PR, so unless we are waiting for @pablog's review I'm gonna ask you to merge it

**Comment by danielaRiesgo** [comment]
🎉
